enhancements/cloud-integration/out-of-tree-provider-support.md

```diff
@@ -401,6 +401,13 @@ Upgrade:
 - `cluster-cloud-controller-manager-operator` starts and creates out-of-tree `CCM` resources for the provider.
 - `machine-config-operator` restart updates the `kubelet` with `--cloud-provider=external` option. At this point the `CCM` is running out-of-tree
 
+Upgrade considerations:
+
+In order to use CCM by default a couple of operators need to have the `library-go` dependency upgraded in a specific order to avoid failures. First, `cluster-cloud-controller-manager-operator` will have library-go updated to enable `CCM`, followed by `cluster-kube-controller-manager-operator`, `machine-config-operator` and then `cluster-kube-apiserver-operator`.
```

Comment:
We can control the order that code lands in git repos, but we can't really control the order that components upgrade in a running cluster. What happens if these components restart during an upgrade in a different sequence than described here? Do the components have some way of knowing that all of their counterparts understand the out-of-tree provider and that it's therefore safe to switch modes?

Reply:
Both upgrades and bootstrap are fine with the current design as long as we can ensure that all components in the target release have the same notion of whether CCM should be enabled.

There are 4 components to consider:
* CCCMO (deploys CCM)
* KCMO (deploys KCM)
* MCO (configures kubelet)
* KASO (deploys apiserver)

I'm going to ignore KASO hereinafter because it's weird and its upgrade order is irrelevant.

CCCMO and KCMO safely negotiate the transfer of the cloud controllers between them, so they can be updated in any order. CCM will be deployed and the cloud controllers disabled in KCM iff both of them are updated.

So there are really only 2 actors here: CCCMO/KCMO and MCO.

If CCCMO/KCMO update before MCO then we will be running CCM in a cluster which is running kubelets using the in-tree cloud provider. The only problem we're aware of in this situation is that Node addresses will be managed both by kubelet and the CCM's node controller. We have a workaround to make this hopefully not too awful in https://github.com/kubernetes/kubernetes/pull/109794. We expect this situation to occur during every upgrade until all kubelets have been updated, which may take a considerable amount of time.

If MCO updates before CCCMO/KCMO and also manages to redeploy some kubelets, we will have kubelets running with the external cloud provider while no external cloud provider is running. The only 2 issues we're aware of in this situation are that node addresses will not be updated and new Nodes cannot be provisioned until CCM is deployed. As soon as CCM is deployed this transitions into the first issue until all kubelets have been updated.

The effects of a partial upgrade are (hopefully) well understood, relatively minor, and temporary as long as the target release is consistent.

If the target release is not consistent the upgrade issues are still manageable, although they will never resolve themselves. However, we have identified a case where bootstrap is broken: i.e. it's possible to upgrade to a partially updated release, but you can't install one from scratch. This would obviously break CI until all patches have landed, and we would like to avoid that. We could also never release in this state.